app/src/androidTest/java/com/karumi/MainActivityTest.kt

```diff
@@ -0,0 +1,156 @@
+    }
+
+    @Test
+    fun showsAvengersBadgeIfASuperHeroIsPartOfTheAvengersTeam() {
+        val superHeroes = givenThereAreSomeAvengers(ANY_NUMBER_OF_SUPER_HEROES)
+
+        startActivity()
+
+        RecyclerViewInteraction.onRecyclerView<SuperHero>(withId(R.id.recycler_view))
+                .withItems(superHeroes)
+                .check({ _, view, e ->
+                    matches(hasDescendant(allOf<View>(withId(R.id.iv_avengers_badge),
+                            withEffectiveVisibility(ViewMatchers.Visibility.VISIBLE)))).check(view, e)
+                })
+    }
+
+    @Test
+    fun doesNotShowAvengersBadgeIfASuperHeroIsNotPartOfTheAvengersTeam() {
+        val superHeroes = givenThereAreSomeSuperHeroes(ANY_NUMBER_OF_SUPER_HEROES, false)
+
+        startActivity()
+
+        RecyclerViewInteraction.onRecyclerView<SuperHero>(withId(R.id.recycler_view))
+                .withItems(superHeroes)
+                .check({ _, view, e ->
+                    matches(hasDescendant(allOf<View>(withId(R.id.iv_avengers_badge),
+                            withEffectiveVisibility(ViewMatchers.Visibility.GONE)))).check(view, e)
+                })
+    }
+
+    @Test
+    fun doesNotShowEmptyCaseIfThereAreSuperHeroes() {
+        givenThereAreSomeSuperHeroes(ANY_NUMBER_OF_SUPER_HEROES)
+
+        startActivity()
+
+        onView(withId(R.id.tv_empty_case)).check(matches(not(isDisplayed())))
+    }
+
+    @Test
+    fun doesNotShowLoadingViewOnceSuperHeroesAreShown() {
+        givenThereAreSomeSuperHeroes(ANY_NUMBER_OF_SUPER_HEROES)
+
+        startActivity()
+
+        onView(withId(R.id.progress_bar)).check(matches(not(isDisplayed())))
+    }
+
+    @Test
+    fun opensSuperHeroDetailActivityOnRecyclerViewItemTapped() {
+        val superHeroes = givenThereAreSomeSuperHeroes()
+        val superHeroIndex = 0
+        startActivity()
+
+        onView(withId(R.id.recycler_view))
+                .perform(actionOnItemAtPosition<RecyclerView.ViewHolder>(superHeroIndex, click()))
+
+        val superHeroSelected = superHeroes[superHeroIndex]
+        intended(hasComponent(SuperHeroDetailActivity::class.java.canonicalName))
+        intended(hasExtra("super_hero_name_key", superHeroSelected.name))
+    }
+
+    @Test
+    fun showsTheExactNumberOfSuperHeroes() {
+        givenThereAreSomeSuperHeroes(ANY_NUMBER_OF_SUPER_HEROES)
+
+        startActivity()
+
+        onView(withId(R.id.recycler_view)).check(
+                matches(recyclerViewHasItemCount(ANY_NUMBER_OF_SUPER_HEROES)))
+    }
+
+    private fun givenThereAreSomeAvengers(numberOfAvengers: Int): List<SuperHero> =
+            givenThereAreSomeSuperHeroes(numberOfAvengers, true)
+
+    private fun givenThereAreSomeSuperHeroes(numberOfSuperHeroes: Int = ANY_NUMBER_OF_SUPER_HEROES, avengers: Boolean = false): List<SuperHero> {
+        val superHeroes = LinkedList<SuperHero>()
+        for (i in 0 until numberOfSuperHeroes) {
```

Comment:
in screenshot superhero we used IntRange(0, numberOfSuperHeroes - 1).map { id -> /**/} instead the for, I don't konw if you llike it or not. I more or less de same, but more kotlin stye.

Reply:
Sure